[TG-1931] Resolve calls to java.lang.Object functions to more specific ones

mgudemann opened the pull request:
needs a public facing regression test

smowton commented:
Please write a commit message explaining the problem and how this fixes it? (Sorry, hit the close button by mistake)

tautschnig commented:
Apart from adding the test, could the commit message please also be made a bit more verbose? Which part of the code is this affecting/where are you gathering calls, and why are you doing this?

thk123 commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -313,17 +309,22 @@ void remove_virtual_functionst::get_child_functions_rec(
 }
 
+/// Used to get dispatch entries to call for the given function
+/// \par function: function that should be called
```
Nit: `param`

thk123 commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -359,17 +360,30 @@ void remove_virtual_functionst::get_functions(
+    functions.begin(),
+    functions.end(),
+    [&concrete_call](const dispatch_table_entryt &entry) {
+      return concrete_call.find(entry.class_id) != concrete_call.end();
```
Maybe being end-of-the-week dense, but rather than first finding the functions to remove then removing them, can this predicate not simply be `return entry.symbol_expr != root_function.symbol_expr`;?

mgudemann replied:
That would also remove those for which the only found `symbol_expr` for the function call is the `root_function.symbol_expr`, wouldn't it? For a class that doesn't implement its own `toString` for example, it makes sense to call `java.lang.Object.toString`. Therefore I collect first entries that do have a more specialized concrete call and remove only those.

But as discussed, it might make sense for example to first have the classes with a more concrete resolved function call and then group those with `root_function.symbol_expr`-"only" into a single conditional to save some GOTO instructions.

For example
```
        // 73 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.StringBuilder" == arg0a->@class_identifier THEN GOTO 3
...
        IF "java::java.util.ArrayList" == arg0a->@class_identifier THEN GOTO 5
        // 58 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.util.Arrays$ArrayList" == arg0a->@class_identifier THEN GOTO 5
...
// all special cases treated, for the rest call `java.lang.Object.toString`
        // 268 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     1: arg0a . java.lang.Object.toString:()Ljava/lang/String;();
        // 269 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.Object.toString:()Ljava/lang/String;#return_value;
...
     3: (struct java.lang.StringBuilder *)arg0a . java.lang.StringBuilder.toString:()Ljava/lang/String;();
        // 277 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.StringBuilder.toString:()Ljava/lang/String;#return_value;
...
        // 284 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     5: (struct java.util.AbstractCollection *)arg0a . java.util.AbstractCollection.toString:()Ljava/lang/String;();
        // 285 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
...
```

tautschnig replied:
> That would also remove those for which the only found `symbol_expr` for the function call is the `root_function.symbol_expr`, wouldn't it? For a class that doesn't implement its own `toString` for example, it makes sense to call `java.lang.Object.toString`. Therefore I collect first entries that do have a more specialized concrete call and remove only those.

I still don't understand: `concrete_calls` is created as a set of keys that are a subset of those contained in `functions` (for some loose notion of keys). Then you remove from `functions` all those that have their key in `concrete_calls`, so you remove an element whenever the previous loop inserted it into `concrete_calls`. Thus it seems that
````c++
std::remove_if(
  functions.begin(),
  functions.end(),
  [](const dispatch_table_entryt &entry) {
    entry.symbol_expr != root_function.symbol_expr;
  });
````
would achieve the same, as @thk123 suggested.

What are we missing?

thk123 reviewed (commented):
As already pointed out needs a test and better commit explanation of why this is an appropriate fix.

mgudemann commented:
@smowton @thk123 @tautschnig I added some explanation in the commit message for what is done in the PR. As so often, there's an internal regression test, but I will try to derive a public one from that.

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -313,17 +309,22 @@ void remove_virtual_functionst::get_child_functions_rec(
   const irep_idt class_id=function.get(ID_C_class);
   const std::string class_id_string(id2string(class_id));
+  // method/function name of function to call
   const irep_idt component_name=function.get(ID_component_name);
```
How about self-documenting code: rename the variable?

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -359,17 +360,32 @@ void remove_virtual_functionst::get_functions(
   if(root_function.symbol_expr!=symbol_exprt())
     functions.push_back(root_function);
+
+  // remove all classes from dispatch table where concrete call has been found
```
I believe this currently filters nothing, as the set contains those entries that resolved to a function other than the root, and the filter condition requires membership of that set *and* having resolved to the root function.

peterschrammel commented:
@mgudemann, is there a TG companion PR for this?

mgudemann commented:
@smowton After removing the `visited` set, there can be multiple entries for one class, e.g., one where the root symbol expression is found and one where a more specialized one is found. The above idea is to record first those where the function call is resolved to something other than root (but potentially also to root) and then removing those from that record that have an entry resolved to root. Therefore something that was, e.g., resolved to `java.lang.Object.toString` and also to `own.package.MyObject.toString` will retain only the second entry.

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -359,17 +360,32 @@ void remove_virtual_functionst::get_functions(
+      return
+        concrete_call.find(entry.class_id) != concrete_call.end() &&
+        entry.symbol_expr == root_function.symbol_expr;
+    });
```
I maintain that I don't understand why this code has to be as complicated.

romainbrenguier replied:
I don't know if it is complicated but it seems to do some unnecessary sorting.
Maybe pull out this code in a function which makes the intent clear, something like `sort_java_lang_object_at_the_end(dispatch_table_entries &)`.
Then it will be easier to see how it can be improved.

smowton commented:
Hmmm. How about this alternate approach:

At the moment we have
```
if(function.symbol_expr == symbol_exprt())
    {
      const resolve_concrete_function_callt::concrete_function_callt
        &resolved_call = resolve_function_call(child, component_name);
```

That takes care of the problem for any method except those implemented by `java.lang.Object` (i.e. `hashCode` and `toString`) which are concrete everywhere. Therefore we could simply handle the case where the current definition is given by j.l.O and treat it the same way.

It's an ugly solution, since it's obviously very Java specific, but so is resolve-concrete-call, which only looks to a class' first parent as a possible source of method implementations. If we want to generalise to C++ or other multiple-inheriting environments (Java 9 default implementations?) then we should do one walk to discover classes of interest, then visit them in topological order, at each juncture taking the most-derived of our parents as the source of the method implementation.

mgudemann commented:
To give you more context @tautschnig.

For the original test case we currently have for the following Java code
```
import java.util.ArrayList;
class ToStringObject {
    public static String test() {
        ArrayList<Integer> al = new ArrayList<>();
        return toStr(al);
    }
    public static String toStr(Object o) {
        return o.toString();
    }
}
```
the following GOTO
```
ToStringObject.toStr() /* java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; */
        // 43 no location
        struct java.lang.String *return_tmp0;
        // 44 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        // Labels: pc0
        ASSERT !(arg0a == null) // Throw null
        // 45 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.Class" == arg0a->@class_identifier THEN GOTO 1
        // 46 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.AbstractStringBuilder" == arg0a->@class_identifier THEN GOTO 2
        // 47 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.util.HashMap$Values" == arg0a->@class_identifier THEN GOTO 4
        // 48 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.util.AbstractCollection" == arg0a->@class_identifier THEN GOTO 4
        // 49 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.StringBuilder" == arg0a->@class_identifier THEN GOTO 5
        // 50 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.String" == arg0a->@class_identifier THEN GOTO 6
        // 51 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.Integer" == arg0a->@class_identifier THEN GOTO 7
        // 52 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        arg0a . java.lang.Object.toString:()Ljava/lang/String;();
        // 53 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.Object.toString:()Ljava/lang/String;#return_value;
        // 54 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        dead java.lang.Object.toString:()Ljava/lang/String;#return_value;
        // 55 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        GOTO 8
...
```

Which means that for `ArrayList.toString` we'd call the `java.lang.Object` implementation, although `java.util.AbstractCollection.toString:()Ljava/lang/String;();` would be the correct one.

in my current PR I get the following for this (with current model loaded, i.e., all possible subclasses of `JLO` somehow appear

```
ToStringObject.toStr() /* java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; */
        // 43 no location
        struct java.lang.String *return_tmp0;
        // 44 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        // Labels: pc0
        ASSERT !(arg0a == null) // Throw null
        // 45 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.AbstractStringBuilder" == arg0a->@class_identifier || "java::java.lang.StringBuffer" == arg0a->@class_identifier THEN GOTO 1
        // 46 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.Class" == arg0a->@class_identifier THEN GOTO 3
        // 47 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.util.ArrayList" == arg0a->@class_identifier || "java::java.util.AbstractList" == arg0a->@class_identifier || "java::java.util.ArrayList$SubList" == arg0a->@class_identifier || "java::java.util.Arrays$ArrayList" == arg0a->@class_identifier || "java::java.util.AbstractCollection" == arg0a->@class_identifier || "java::java.util.HashMap$KeySet" == arg0a->@class_identifier || "java::java.util.HashMap$Values" == arg0a->@class_identifier || "java::java.util.HashMap$EntrySet" == arg0a->@class_identifier || "java::java.util.AbstractSet" == arg0a->@class_identifier THEN GOTO 4
        // 48 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.Integer" == arg0a->@class_identifier THEN GOTO 5
        // 49 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.String" == arg0a->@class_identifier THEN GOTO 6
        // 50 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::java.lang.StringBuilder" == arg0a->@class_identifier THEN GOTO 7
        // 51 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        arg0a . java.lang.Object.toString:()Ljava/lang/String;();
        // 52 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.Object.toString:()Ljava/lang/String;#return_value;
        // 53 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        dead java.lang.Object.toString:()Ljava/lang/String;#return_value;
        // 54 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        GOTO 8
        // 55 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     1: (struct java.lang.AbstractStringBuilder *)arg0a . java.lang.AbstractStringBuilder.toString:()Ljava/lang/String;();
        // 56 no location
        IF !NONDET(_Bool) THEN GOTO 2
        // 57 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = null;
        // 58 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     2: GOTO 8
        // 59 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     3: (struct java.lang.Class *)arg0a . java.lang.Class.toString:()Ljava/lang/String;();
        // 60 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.Class.toString:()Ljava/lang/String;#return_value;
        // 61 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        dead java.lang.Class.toString:()Ljava/lang/String;#return_value;
        // 62 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        GOTO 8
        // 63 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     4: (struct java.util.AbstractCollection *)arg0a . java.util.AbstractCollection.toString:()Ljava/lang/String;();
        // 64 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.util.AbstractCollection.toString:()Ljava/lang/String;#return_value;
        // 65 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        dead java.util.AbstractCollection.toString:()Ljava/lang/String;#return_value;
        // 66 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        GOTO 8
        // 67 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     5: (struct java.lang.Integer *)arg0a . java.lang.Integer.toString:()Ljava/lang/String;();
        // 68 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.Integer.toString:()Ljava/lang/String;#return_value;
        // 69 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        dead java.lang.Integer.toString:()Ljava/lang/String;#return_value;
        // 70 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        GOTO 8
        // 71 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     6: (struct java.lang.String *)arg0a . java.lang.String.toString:()Ljava/lang/String;();
        // 72 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.String.toString:()Ljava/lang/String;#return_value;
        // 73 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        dead java.lang.String.toString:()Ljava/lang/String;#return_value;
        // 74 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        GOTO 8
        // 75 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     7: (struct java.lang.StringBuilder *)arg0a . java.lang.StringBuilder.toString:()Ljava/lang/String;();
        // 76 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = java.lang.StringBuilder.toString:()Ljava/lang/String;#return_value;
        // 77 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        dead java.lang.StringBuilder.toString:()Ljava/lang/String;#return_value;
        // 78 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 2
     8: ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String;#return_value = return_tmp0;
        // 79 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 2
        dead return_tmp0;
        // 80 no location
        END_FUNCTION
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
This groups together the `class_id`s with the same call target, ending with a call to `java.lang.Object` which is simply the last, non-explicit case.

Do we agree that at least this result is more or less what we want to have? Is my assumption correct that having larger disjunctions instead of guarded GOTO is beneficial from a performance point of view? In any case it should be easier on the `--depth` parameter if many such methods are encountered, right?

mgudemann commented:
@smowton not sure that I understand you here correctly. The problem here at hand is that the original implementation tries to optimize which classes it visits which can be problematic as shown in the above example and the JIRA issue.

The problem is not that `symbol_expr` is empty but that it contains the wrong function to call. As I see it, if we have a function with parameter `T t` then we'll need to give an explicit function call for each possible subclass of type `T`. For JLO this effectively means having to enumerate all known object types.

smowton commented:
What I mean is, currently AFAIK we re-visit a type if we don't have a resolution for it yet (i.e. we've previously seen it via an interface, but might be about to visit it via a concrete supertype). We could simply expand that logic to say re-visit also if the current resolution is to java.lang.Object.*

mgudemann commented:
@smowton imagine the method doesn't take an JLO parameter, but some other object type `T`, if there's a subclass of `T` which overrides a function, we'd still like to use the more specialized version. Or is this case not possible ?

smowton commented:
The only case where we can encounter interfaces while walking child types is if our starting point was either itself an interface, or is java.lang.Object. Arguably that means we should make interfaces not children of j.l.O! But in general if we started from a concrete type then our only children are also concrete types and the child graph is a tree, so this bug doesn't arise. The only case where it does is when either we start at an interface (with no definitions) or with Object (with toString and hashCode definitions)

mgudemann commented:
In any case, my main question now is whether the result as above is what we want to have. How we get there in the end is a separate thing and I agree that it can be done in a simpler way than what I currently propose.
 @smowton it could be an abstract class with an implementation for the method in question, too

Commit:
Adapt regression tests for virtual functions.

These tests were failing when the new way was used which groups class
identifiers that have the same resolved function call.

Unfortunately the result is not unique here which makes those tests not too
useful.

smowton commented:
If it's an abstract class (but not an interface) then its child class graph is again a tree.

That GOTO program looks good except for the label `2:`, which looks like an empty callee block got written out? It's harmless in this program but potentially concerning.

mgudemann commented:
@smowton maybe an issue with String handling somehow (this is for StringBuilder), I think this is the same in the current implementation.

it returns `null`, depending on the non-det Boolean in 56

```
        // 55 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     1: (struct java.lang.AbstractStringBuilder *)arg0a . java.lang.AbstractStringBuilder.toString:()Ljava/lang/String;();
        // 56 no location
        IF !NONDET(_Bool) THEN GOTO 2
        // 57 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        return_tmp0 = null;
        // 58 file ToStringObject.java line 9 function java::ToStringObject.toStr:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
     2: GOTO 8
```

smowton commented:
How odd, not sure what's going on there, but disregard my speculation that this had anything to do with remove-returns :)

mgudemann commented:
@smowton I'll try your suggestion, I think we should take the simplest solution we can find.

Commit:
Re-Resolve function calls if only java.lang.Object was found.

This fixes an issue in test-generator where a call to `toString` is dispatched
to the `toString` method from `java.lang.Object` is used instead of the
overridden method from `ArrayList`.

Cite from the issue about the current implementation:

> It tracks a `visited` set preventing it from listing callees twice when
> multiple inheritance is in play (e.g. Java interfaces). Unfortunately this
> malfunctions when we visit a type twice, once via its interfaces and once via
> its concrete subclass which provides a definition

This extends the original implementation, by resolving dispatch entries where an
initial step resolved to a java.lang.Object function. This case can be an error,
as some classes might be visited multiple times, first for an interface and then
for the concrete class.

The original implementation kept a visited set that recorded visited classes
which resulted in some functions not correctly being resolved. This set is
replaced with a map from class identifiers to dispatch table entries.

Commit:
Add regression test

In the original implementation, D.toString would be resolved to
java.lang.Object.toString().

mgudemann commented:
@smowton @tautschnig I simplified this with an approximation to what @smowton suggested. I also added a public regression test that shows the problem.

thk123 commented on regression/cbmc-java/virtual10/test.java:
```diff
@@ -0,0 +1,33 @@
+interface A {
+  public int f();
```
What is the importance of these methods? Seems like `Object` can only call `O::toString` or `C::toString`?

thk123 reviewed (commented):
Don't fully understand the relevance of the regression test here?

mgudemann commented:
@thk123 it shows that "D" is treated correctly now, in the sense that O.toString() is called. In the current implementation it fell back to java.lang.Object.toString().

The interfaces (one would probably suffice) make this a pseudo multi-inheritance case which leads to the problem, as a class can be visited as child of the interface and have its call resolved to Object.toString and also be visited as sub-class of Object and have its function call resolved to something else.

before
```
        IF "java::O" == o->@class_identifier THEN GOTO 2
        // 44 file test.java line 32 function java::E.f:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::C" == o->@class_identifier THEN GOTO 3
        // 45 file test.java line 32 function java::E.f:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        o . java.lang.Object.toString:()Ljava/lang/String;();
```
with this PR
```
        // 43 file test.java line 32 function java::E.f:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::D" == o->@class_identifier || "java::O" == o->@class_identifier THEN GOTO 2
        // 44 file test.java line 32 function java::E.f:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        IF "java::C" == o->@class_identifier THEN GOTO 3
        // 45 file test.java line 32 function java::E.f:(Ljava/lang/Object;)Ljava/lang/String; bytecode-index 1
        o . java.lang.Object.toString:()Ljava/lang/String;();
        // 46 no location
```

Commit:
Add unit test

The unit test checks whether the correct function is called. It is therefor more
precise than the regression test `virtual10`.

mgudemann commented:
@smowton Would be ready for a hopefully final review. I added a unit test that checks for correct target function calls.

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -209,15 +216,54 @@ void remove_virtual_functionst::remove_virtual_function(
+    // Emit target if end of dispatch table is reached or if the next element is
+    // dispatched to another function call. Assumes entries in the functions
+    // variable to be sorted for the identifier of the function to be called.
+    auto l_it = it;
```
` = std::next(it)`

smowton commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -209,15 +216,54 @@ void remove_virtual_functionst::remove_virtual_function(
+    // variable to be sorted for the identifier of the function to be called.
+    auto l_it = it;
+    l_it++;
+    bool next_emit_target = l_it == functions.crend();
```
Probably clearer as a straightforward `||`

smowton reviewed (approved):
Looks good, couple of nitpicks.

tautschnig commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -209,15 +216,54 @@ void remove_virtual_functionst::remove_virtual_function(
+    bool next_emit_target = l_it == functions.crend();
+    if(!next_emit_target)
+    {
+      next_emit_target |=
```
You know that `next_emit_target` is false, so `|=` is the same as `=` - seems fishy?

romainbrenguier commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -209,15 +216,53 @@ void remove_virtual_functionst::remove_virtual_function(
+    {
+      next_emit_target =
+        l_it->symbol_expr.get_identifier() != fun.symbol_expr.get_identifier();
+    }
```
these 6 lines could be rewriten `bool next_emit_target = (l_it == functions.crend()) || (l_it->symbol_expr.get_identifier() != fun.symbol_expr.get_identifier();`

romainbrenguier commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -209,15 +216,53 @@ void remove_virtual_functionst::remove_virtual_function(
+        exprt::operandst or_ops;
+        for(const auto &id : class_ids)
+        {
+          exprt c_id1 = constant_exprt(id, string_typet());
```
`const constant_exprt c_id1(id, string_typet());`

romainbrenguier commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -209,15 +216,53 @@ void remove_virtual_functionst::remove_virtual_function(
+        for(const auto &id : class_ids)
+        {
+          exprt c_id1 = constant_exprt(id, string_typet());
+          equal_exprt class_id_test(c_id1, c_id2);
```
const

romainbrenguier commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -209,15 +216,53 @@ void remove_virtual_functionst::remove_virtual_function(
+
+        goto_programt::targett t4 = new_code_gotos.add_instruction();
+        t4->source_location = vcall_source_loc;
+        t4->make_goto(insertit.first->second, disjunction(or_ops));
```
I find the variable name `t4` not very descriptive, I would suggest something like `target_for_goto_instruction`.

romainbrenguier commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -357,17 +418,36 @@ void remove_virtual_functionst::get_functions(
   if(root_function.symbol_expr!=symbol_exprt())
     functions.push_back(root_function);
+
+  // Sort for call symbol expr grouping, keep java.lang.Object entries at the
```
not sure how to parse that

romainbrenguier commented on src/goto-programs/remove_virtual_functions.cpp:
```diff
@@ -357,17 +418,36 @@ void remove_virtual_functionst::get_functions(
     functions.push_back(root_function);
+
+  // Sort for call symbol expr grouping, keep java.lang.Object entries at the
+  // end for fall through. The reasoning is that this is case with most entries
```
*the* case

Commit:
Adapt failing unit test for value set analysis

The test was failing on two Travis goals, due to the config object being
persistent between different tests in "unit". The link order made another test
set `config.main` to non-empty which made this test fail due to that function
not being found.
